src/main/java/ch/njol/skript/lang/function/DefaultFunction.java

```diff
@@ -0,0 +1,341 @@
+					arg[0].getClass().getSimpleName());
+
+				args.put(parameter.getName(), arg[0]);
+			} else {
+				assert parameter.getType().getC().isAssignableFrom(arg.getClass())
+					: "argument type %s does not match parameter type %s".formatted(parameter.getType().getC().getSimpleName(),
+					arg.getClass().getSimpleName());
+
+				args.put(parameter.getName(), arg);
+			}
+		}
+
+		FunctionArguments arguments = new FunctionArguments(args);
+		T result = execute.apply(arguments);
+
+		if (result == null) {
+			return null;
+		} else if (result.getClass().isArray()) {
+			//noinspection unchecked
+			return (T[]) result;
+		} else {
+			//noinspection unchecked
+			T[] array = (T[]) Array.newInstance(result.getClass(), 1);
+			array[0] = result;
+			return array;
+		}
+	}
+
+	@Override
+	public boolean resetReturnValue() {
+		return true;
+	}
+
+	/**
+	 * Returns this function's description.
+	 *
+	 * @return The description.
+	 */
+	public @NotNull String[] description() {
+		return description;
+	}
+
+	/**
+	 * Returns this function's version history.
+	 *
+	 * @return The version history.
+	 */
+	public @NotNull String[] since() {
+		return since;
+	}
+
+	/**
+	 * Returns this function's examples.
+	 *
+	 * @return The examples.
+	 */
+	public @NotNull String[] examples() {
+		return examples;
+	}
+
+	/**
+	 * Returns this function's keywords.
+	 *
+	 * @return The keywords.
+	 */
+	public @NotNull String[] keywords() {
+		return keywords;
+	}
+
+	/**
+	 * Registers this function.
+	 *
+	 * @return This function.
+	 * @see Functions#register(DefaultFunction)
+	 */
+	@Contract(" -> this")
+	public DefaultFunction<T> register() {
+		Functions.register(this);
+
+		return this;
+	}
+
+	public static class Builder<T> {
+
+		private final String name;
+		private final Class<T> returnType;
+		private final Map<String, Parameter<?>> parameters = new LinkedHashMap<>();
+
+		private ch.njol.skript.util.Contract contract = null;
+
+		private String[] description, since, examples, keywords;
+
+		private Builder(@NotNull String name, @NotNull Class<T> returnType) {
+			Preconditions.checkNotNull(name, "name cannot be null");
+			Preconditions.checkNotNull(returnType, "return type cannot be null");
+
+			this.name = name;
+			this.returnType = returnType;
+		}
+
+		/**
+		 * Sets this function builder's {@link Contract}.
+		 *
+		 * @param contract The contract.
+		 * @return This builder.
+		 */
+		@Contract("_ -> this")
+		public Builder<T> contract(@NotNull ch.njol.skript.util.Contract contract) {
+			Preconditions.checkNotNull(contract, "contract cannot be null");
+
+			this.contract = contract;
+			return this;
+		}
+
+		/**
+		 * Sets this function builder's description.
+		 *
+		 * @param description The description.
+		 * @return This builder.
+		 */
+		@Contract("_ -> this")
+		public Builder<T> description(@NotNull String... description) {
+			Preconditions.checkNotNull(description, "description cannot be null");
```

Comment:
The `Preconditions` check ensures that the array itself is not null, whereas you're marking the strings within the array as not null. This would be more accurate:
```suggestion
		public Builder<T> description(String @NotNull ... description) {
			Preconditions.checkNotNull(description, "description cannot be null");
```